graph/src/components/subgraph/proof_of_indexing/online.rs

```diff
@@ -54,82 +92,75 @@ pub enum BlockEventStream {
-            1,                                // kvp -> v
-            0,                                // CausalityRegion.blocks: Vec<Block>
-            block_number.try_into().unwrap(), // Vec<Block> -> [i]
-            0,                                // Block.events -> Vec<ProofOfIndexingEvent>
-        ]);
-        Self::LegacyHasherStream {
+    fn new(block_number: BlockNumber, version: ProofOfIndexingVersion) -> Self {
+        let block_index: u64 = block_number.try_into().unwrap();
+
+        Self {
             vec_length: 0,
             handler_start: 0,
-            seq_no: events,
-            digest: SetHasher::new(),
+            block_index,
+            hasher: Hashers::new(version),
         }
     }
 
     /// Finishes the current block and returns the serialized hash function to
     /// be resumed later. Cases in which the hash function is resumed include
     /// when asking for the final PoI, or when combining with the next modified
     /// block via the argument `prev`
-    pub fn pause(self, prev: Option<&[u8]>) -> Vec<u8> {
-        match self {
-            BlockEventStream::LegacyHasherStream {
-                vec_length,
-                handler_start: _,
-                seq_no,
-                mut digest,
-            } => {
-                vec_length.stable_hash(seq_no, &mut digest);
-                let mut state = digest;
+    pub fn pause(mut self, prev: Option<&[u8]>) -> Vec<u8> {
+        self.hasher
+            .write(&self.vec_length, &[1, 0, self.block_index, 0]);
+        match self.hasher {
+            Hashers::Legacy(mut digest) => {
                 if let Some(prev) = prev {
                     let prev = SetHasher::from_bytes(prev);
-                    state.finish_unordered(prev, SequenceNumber::root());
+                    // SequenceNumber::root() is misleading here since the parameter
+                    // is unused.
+                    digest.finish_unordered(prev, SequenceNumber::root());
                 }
-                state.to_bytes()
+                digest.to_bytes()
+            }
+            Hashers::Fast(mut digest) => {
+                if let Some(prev) = prev {
+                    let prev = prev
+                        .try_into()
+                        .expect("Expected valid fast stable hash representation");
+                    let prev = FastStableHasher::from_bytes(prev);
+                    digest.mixin(&prev);
+                }
+                digest.to_bytes().to_vec()
             }
         }
     }
 
     fn write(&mut self, event: &ProofOfIndexingEvent<'_>) {
-        match self {
-            BlockEventStream::LegacyHasherStream {
-                ref mut vec_length,
-                handler_start: _,
-                seq_no,
-                ref mut digest,
-            } => {
-                *vec_length += 1;
-                event.stable_hash(seq_no.next_child(), digest);
-            }
-        }
+        let children = &[
```

Comment:
Could use some more information about what is going on here. I might just be missing context but overall this section is not very clear to me even with the existing comments.